### PR TITLE
generate device context managers in inductor code

### LINKED ISSUE #1717: [Bug]: Inductor IMA's when inputs are on a non-default dvice
### 🐛 Describe the bug

-

### Error logs

cuda illegal memory access

### Did Dynamo succeed?

- [ ] Does dynamo.optimize("eager") succeed?

### Did AOT succeed?

- [ ] Did dynamo.optimize("aot_eager") succeed?

### Did Inductor succeed?

- [ ] Does dynamo.optimize("inductor") succeed?

### Minified repro

```
import torch
import torch._inductor
import torch._inductor.config as config

#torch._dynamo.config.dynamic_shapes = True
torch._inductor.config.triton.cudagraphs = False
torch._inductor.config.debug=True
import torch._dynamo as torchdynamo
@torchdynamo.optimize("inductor")
def div(x, y):
    return torch.ops.aten.div(x, y)


x = torch.randn(4, dtype=torch.float, device="cuda:1")
y = torch.randn(4, dtype=torch.float, device="cuda:1")
print(div(x,y))
```
In eager we set device guard to temporarily set current device to the device of the inputs, inductor doesn't do that.

### PR DESCRIPTION
**Pull request by ngimel**
Fixes https://github.com/pytorch/torchdynamo/issues/1717, https://github.com/pytorch/pytorch/issues/93478

<s>TODO: add test with multiple devices, figure out extra context initialization</s>

Problems:
<s>It still initializes context on 0-th device that it shouldn't, I'll take a look where that happens and fix before landing</s>
It adds a python device context manages, that is absurdly slow and takes ~2.5 us (should be nanoseconds). That's not a problem for real models, because it'll be called just once, but it is a bit of an inconvenience for microbenchmarking, we should make that context manager more performant (won't fix in this PR)
It still can have bugs for graphs that run on multiple devices and can have buffers incorrectly shared between multiple device by memory reuse, if that happens that'll need to be solved separately.  

Generated code:
```
def call(args):
    arg0_1, arg1_1 = args
    args.clear()
    with torch.cuda.device(1):
        buf0 = empty_strided((4, ), (1, ), device='cuda', dtype=torch.float32)
        stream1 = get_cuda_stream(1)
        triton_fused_div_0.run(arg0_1, arg1_1, buf0, 4, grid=grid(4), stream=stream1)
        del arg0_1
        del arg1_1
        return (buf0, )
```

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: torch/_inductor/triton_ops/autotune.py [hunk 1]
```diff
@@ -80,11 +79,13 @@ def _precompile_config(self, cfg: Config, warm_cache_only_with_cc: int):
             return
 
         torch.cuda.set_device(torch.cuda.current_device())
-
-        binary = triton.compile(
-            self.fn,
-            **compile_meta,
-        )
+        with torch.cuda.device(compile_meta["device"]):
```
Comment (wconstab) [inline torch/_inductor/triton_ops/autotune.py:82]: is the set_device still needed on L81 above?

File: torch/_inductor/triton_ops/autotune.py [hunk 2]
```diff
@@ -80,11 +79,13 @@ def _precompile_config(self, cfg: Config, warm_cache_only_with_cc: int):
...
+            # need to initialize context
+            torch.cuda.synchronize(torch.cuda.current_device())
```
Comment (wconstab) [inline torch/_inductor/triton_ops/autotune.py:84]: why is this needed?

File: torch/_inductor/triton_ops/autotune.py (hunk 1 above)
Comment (ngimel) [reply torch/_inductor/triton_ops/autotune.py:82]: Right, that set_device is not needed, thanks!

File: torch/_inductor/triton_ops/autotune.py (hunk 2 above)
Comment (ngimel) [reply torch/_inductor/triton_ops/autotune.py:84]: Otherwise `triton.compile` errors out with "context not initialized", which should properly be fixed on triton side, but this will do for now.

**Review by wconstab — APPROVED** [review]
lgtm, thanks for fixing!

File: torch/_inductor/triton_ops/autotune.py (hunk 2 above)
Comment (pritamdamania87) [reply torch/_inductor/triton_ops/autotune.py:84]: @wconstab @ngimel Are we referring to the cuda context here or some other context? If so, I'm assuming it should be possible to first check if the context is initialized and only do this if the context is not initialized? See https://github.com/pytorch/pytorch/issues/109074 for context
